_source/_posts/2022-04-11-java-cli-device-grant.adoc

```diff
@@ -0,0 +1,412 @@
+
+* You should ONLY enter a password into the website it was created for.
+* Your terminal doesn't support the same factors your browser does.
+* You may be ssh'd into another computer.
+
+NOTE: The Authorization Device Grant requires a user. If you are trying to call APIs from an automated process or other server-to-server communication, check out the link:/blog/2021/05/05/client-credentials-spring-security[OAuth 2.0 Client Credential Flow].
+
+== Setup Okta
+
+First, you need to create an Okta application that supports the device grant.
+
+. In **Applications**->**Applications**, click **Create App Integration**. In the pop-up, choose **OIDC - OpenID Connect**, then **Native Application**. The Device Grant is only available for Native Applications.
++
+image::{% asset_path 'blog/java-cli-device-grant/create-native-app.jpg' %}[alt=Set up a native app,width=800,align=center]
+
+. In the next step, check **Device Authorization** in the **Grant type** section.
++
+image::{% asset_path 'blog/java-cli-device-grant/enable-device-flow.jpg' %}[alt=Enable device grant,width=800,align=center]
+
+Make a note of the **Client ID**, you will need it below.
+
+If you are using an https://developer.okta.com/docs/concepts/auth-servers/#available-authorization-server-types[Okta Authorization Server], you also need to enable the grant type in the authorization server's policy interface:
+
+. Go to **Security**->**API**
+. Select the authorization server you want to use (for example, "default").
+. Click the **Access Policies** tab, and edit the rule for your policy. Make sure to select the **Device Authorization** check box.
++
+image::{% asset_path 'blog/java-cli-device-grant/custom-as.jpg' %}[alt=Enable device flow on a custom Authorization Server,width=800,align=center]
+
+Make note of your Issuer URI, if you are using an Okta Custom Authorization Server, it will look something like this (updating the hostname to match your Okta Org):
+
+[source,txt]
+----
+https://dev-133337.okta.com/api/v1/authorizationServers/default
+----
+
+Otherwise, it will be the same as your Okta Url:
+
+[source,txt]
+----
+https://dev-133337.okta.com
+----
+
+== Create a Java application
+
+On to the fun part, writing some code!
+
+Create a new project in your favorite IDE (I'm a fan of https://www.jetbrains.com/idea/[IntelliJ IDEA]). The code below uses the `java.net.http.HttpClient`, make sure our target Java is set to at least `11`.
+
+The `HttpClient` API is much easier to use than its predecessor `URLConnection`, but Java is still no way to parse JSON out of the box. Add a dependency on https://github.com/FasterXML/jackson-databind[Jackson Databind] to fix that.
+
+To keep things all contained in a single `java` file, I'll use https://www.jbang.dev[JBang]. If you haven't used JBang before, the TL;DR it makes it easy to create Java... scripts. Your project dependencies get added to the top of the `java` file.
```

Comment:
```suggestion
To keep things all contained in a single `java` file, I'll use https://www.jbang.dev[JBang]. If you haven't used JBang before, here's the TL;DR: JBang makes it easy to create Java... scripts. It's a scripting tool that  lets you run Java application with minimal setup. Your project dependencies get added to the top of the `java` file.
```
I love the pun, but i added a wee bit more explanation. Pls check that it's accurate.